common.rkt

```diff
@@ -150,17 +145,14 @@
                     ((var? t)  (set! index (+ 1 index))
                                (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
                     (else      st)))))
-    (cond
-      ((and (null? (state-diseq st)) (null? (state-types st))) (walk* tm results))
-      ((null? (state-types st)) (Ans 
-                                  (walk* tm results)
-                                  (list (walk* (cons '=/= (map pretty-diseq (state-diseq st))) results))))
-      ((null? (state-diseq st)) (Ans
-                                  (walk* tm results)
-                                  (walk* (map pretty-types (state-types st)) results)))
-      (else (Ans (walk* tm results) 
-                (append (walk* (map pretty-types (state-types st)) results) 
-                        (list (walk* (cons '=/= (map pretty-diseq (state-diseq st))) results) )))))))
+    (let* ((walked-sub (walk* tm results))
+          (diseq-null (null? (state-diseq st)))
+          (types-null (null? (state-types st)))
+          (types (walk* (map pretty-types (state-types st)) results))
+          (diseq (if diseq-null '() (list (walk* (cons '=/= (map pretty-diseq (state-diseq st))) results)))))
+      (if (and diseq-null types-null)
+          walked-sub
+          (Ans walked-sub (append types diseq))))))
```

Comment:
I'd write this a bit differently, both to be a little more straightforward about what we're trying to distinguish, and to be a little more future proof in case we add more kinds of constraints later.

First, I would give a name to `(append types diseq)` in the `let*` above.  Let's call it `cxs`.

Then I would change the condition here from `(and diseq-null? types-null?)` to `(null? cxs)`.  And of course also use it in `(Ans walked-sub cxs)`.

This also means you don't really need to give names to `diseq-null?` and `types-null?`, so I would inline those in the one place they're used.